libqtile/layout/plasma.py

```diff
@@ -0,0 +1,1085 @@
+            return self.first_leaf
+        idx = self.index + 1
+        if idx >= len(self.parent):
+            return self.parent.next_leaf
+        return self.parent[idx].first_leaf
+
+    @property
+    def all_leafs(self):
+        if self.is_leaf:
+            yield self
+        for child in self:
+            yield from child.all_leafs
+
+    @property
+    def orient(self):
+        if self.is_root:
+            return self.root_orient
+        return ~self.parent.orient
+
+    @property
+    def horizontal(self):
+        return self.orient is Orient.HORIZONTAL
+
+    @property
+    def vertical(self):
+        return self.orient is Orient.VERTICAL
+
+    @property
+    def x(self):
+        if self.is_root:
+            return self._x
+        if self.horizontal:
+            return self.parent.x
+        return self.parent.x + self.size_offset
+
+    @x.setter
+    def x(self, val):
+        if not self.is_root:
+            return
+        self._x = val
+
+    @property
+    def y(self):
+        if self.is_root:
+            return self._y
+        if self.vertical:
+            return self.parent.y
+        return self.parent.y + self.size_offset
+
+    @y.setter
+    def y(self, val):
+        if not self.is_root:
+            return
+        self._y = val
+
+    @property
+    def pos(self):
+        return Point(self.x, self.y)
+
+    @property
+    def width(self):
+        if self.is_root:
+            return self._width
+        if self.horizontal:
+            return self.parent.width
+        return self.size
+
+    @width.setter
+    def width(self, val):
+        if self.is_root:
+            self._width = val
+        elif self.horizontal:
+            self.parent.size = val
+        else:
+            self.size = val
+
+    @property
+    def height(self):
+        if self.is_root:
+            return self._height
+        if self.vertical:
+            return self.parent.height
+        return self.size
+
+    @height.setter
+    def height(self, val):
+        if self.is_root:
+            self._height = val
+        elif self.vertical:
+            self.parent.size = val
+        else:
+            self.size = val
+
+    @property
+    def x_end(self):
+        return self.x + self.width
+
+    @property
+    def y_end(self):
+        return self.y + self.height
+
+    @property
+    def x_center(self):
+        return self.x + self.width / 2
+
+    @property
+    def y_center(self):
+        return self.y + self.height / 2
+
+    @property
+    def center(self):
+        return Point(self.x_center, self.y_center)
+
+    @property
+    def top_left(self):
+        return Point(self.x, self.y)
+
+    @property
+    def top_right(self):
+        return Point(self.x + self.width, self.y)
+
+    @property
+    def bottom_left(self):
+        return Point(self.x, self.y + self.height)
+
+    @property
+    def bottom_right(self):
+        return Point(self.x + self.width, self.y + self.height)
```

Comment:
these are only here for testing code (and probs other properties too), we can just e.g. self.x and self.y + self.height in the test itself

Reply:
Agreed.